src/main/kotlin/org/elm/lang/core/types/TypeReplacement.kt

```diff
@@ -77,38 +71,63 @@ class TypeReplacement(
+         * Use this to prevent cached values from being altered. Note that this is an in-place
+         * change rather than a copy.
          */
-        fun freeze(ty: Ty): Ty {
-            if (ty.traverse(true).none { it.isUnfrozenRecord() }) return ty
-            return TypeReplacement(emptyMap(), freshen = false, varsToRemainRigid = emptyList(), freeze = true, keepRecordsMutable = false).replace(ty)
-        }
-
-        private fun tyWouldChange(ty: Ty, freeze: Boolean): Boolean {
-            return ty.traverse(true).any {
-                it is TyVar || freeze && it.isUnfrozenRecord()
+        fun freeze(ty: Ty) {
+            when (ty) {
+                is TyVar -> Unit
+                is TyTuple -> ty.types.forEach { freeze(it) }
+                is TyRecord -> {
+                    (ty.baseTy as? TyRecord)?.fieldReferences?.freeze()
+                    ty.fields.values.forEach { freeze(it) }
+                    ty.fieldReferences.freeze()
+                }
+                is MutableTyRecord -> {
+                    (ty.baseTy as? TyRecord)?.fieldReferences?.freeze()
+                    ty.fields.values.forEach { freeze(it) }
+                    ty.fieldReferences.freeze()
+                }
+                is TyUnion -> ty.parameters.forEach { freeze(it) }
+                is TyFunction -> {
+                    freeze(ty.ret)
+                    ty.parameters.forEach { freeze(it) }
+                }
+                is TyUnit -> Unit
+                is TyUnknown -> Unit
+                TyInProgressBinding -> Unit
             }
+            ty.alias?.parameters?.forEach { freeze(it) }
         }
+    }
 
-        private fun Ty.isUnfrozenRecord() = when (this) {
-            is TyRecord -> !fieldReferences.frozen
-            is MutableTyRecord -> !fieldReferences.frozen
-            else -> false
-        }
+    private fun wouldChange(ty: Ty): Boolean {
+        val changeAllVars = freshen || varsToRemainRigid != null
+        fun Ty.f(): Boolean = when (this) {
+            is TyVar -> changeAllVars || this in replacements
+            is TyTuple -> types.any { it.f() }
+            is TyRecord -> fields.values.any { it.f() } || baseTy?.f() == true
+            is MutableTyRecord -> !keepRecordsMutable || fields.values.any { it.f() } || baseTy?.f() == true
+            is TyUnion -> parameters.any { it.f() }
+            is TyFunction -> ret.f() || parameters.any { it.f() }
+            is TyUnit, is TyUnknown, TyInProgressBinding -> false
+        } || alias?.parameters?.any { it.f() } == true
+        return ty.f()
     }
 
     /** A map of var to (has been accessed, ty) */
     private val replacements = replacements.mapValuesTo(mutableMapOf()) { (_, v) -> false to v }
 
     fun replace(ty: Ty): Ty {
-        val replaced = when (ty) {
+        // If we wouldn't change anything, return the original ty to avoid duplicating the object
+        if (!wouldChange(ty)) return ty
```

Comment:
Besides the obvious improvement of not performing a copy then throwing it away if it was equal to the original, this also lets of get rid of using ty equality at all. I never liked using equality, since it was the source of a couple of hard to track down bugs.